### PR TITLE
Stdlib: Actor.bt named-registration API (BT-1988)

### PR DESCRIPTION
**Pull request by jamesc**
Phase 2 of [ADR 0079](https://github.com/jamesc/beamtalk/blob/main/docs/ADR/0079-named-actor-registration.md) — adds the user-facing named-registration API to `Actor.bt`, backed by the runtime intrinsics from BT-1987.

Linear: [BT-1988](https://linear.app/beamtalk/issue/BT-1988)

## Summary

- New `Actor` API: `class spawnAs:`, `class spawnWith:as:`, `registerAs:`, `unregister`, `registeredName`, `isRegistered`, `class named:`, `class allRegistered`.
- Return types use `Result(Self, Error)` per ADR 0079 + BT-1986 so subclass lookups narrow correctly (`Counter named: #c` → `Result(Counter, Error)`).
- Codegen now injects the `'$beamtalk_actor' => ClassName` process-dictionary marker into every generated actor `init/1`. BT-1987 only set it for actors that delegated to `beamtalk_actor:init/1`; this PR plumbs it through compiled actors too.
- The marker write is guarded by `__skip_initialize__` so a parent class called as a state-building helper does not overwrite the leaf subclass's marker.

## Key changes

- `stdlib/src/Actor.bt` — 8 new methods (class-side + instance-side), all `sealed`. `unregister` discards the FFI-converted `Result ok: nil` and returns `#ok` per ADR 0079's Symbol contract.
- `runtime/apps/beamtalk_runtime/src/beamtalk_actor.erl` — FFI shims `doSpawnAs/2`, `doSpawnWith/3`, `registerAs/2`, `unregister/1`, `registeredName/1`, `isRegistered/1`, `named/2`, `allRegistered/1`. Deliberately different from the runtime `spawnAs/2,3` entry points to avoid FFI dispatch collision.
- `runtime/apps/beamtalk_runtime/src/beamtalk_behaviour_intrinsics.erl` — exports `walk_hierarchy/3` so `named:` can do the subclass check without duplicating traversal logic.
- `crates/beamtalk-core/src/codegen/core_erlang/gen_server/callbacks.rs` — emits the guarded `erlang:put('$beamtalk_actor', Class)` at the top of every actor `init/1`.

## Tests

- **BUnit** (`stdlib/test/actor_named_test.bt`, 17 tests): success paths for each method, duplicate registration, reserved-name rejection, wrong-class lookup, `unregister` idempotency on registered + unregistered receivers, `registeredName`/`isRegistered` on both states, `allRegistered` round-trip, OTP-kernel filter, and a **subclass-marker regression guard** that exercises `InheritingCounter` → `Counter` and verifies lookups succeed on both the leaf and parent class.
- **EUnit** (`beamtalk_actor_tests.erl`): type-error edge cases for each shim (non-`beamtalk_object` receivers, non-atom names, bogus class-self).
- All 1844 BUnit, 3444 runtime EUnit, and 1242 workspace EUnit tests pass.

## Out of scope

- `SupervisionSpec withName:` — separate issue (BT-1989, same phase).
- Proxy send-site dispatch / name re-resolution across restarts — Phase 3 (BT-1990).

## Test plan

- [x] `just test-bunit` passes
- [x] `just test` (runtime EUnit) passes
- [x] `just clippy` clean
- [x] `just ci` (modulo CodeRabbit review handling)
- [ ] Manual REPL session: `Counter spawnAs: #c` → `Counter named: #c` (deferred to Phase 3 REPL integration)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  - Added Actor named-registration APIs: class-side spawnAs:/spawnWith:as:, lookup named:/allRegistered, and instance-side registerAs:/unregister/registeredName/isRegistered; ensures subclass lookup works correctly after registration.

* **Tests**
  - Added end-to-end test suite and fixtures (NamedCounter, NamedGreeter) covering success, error cases, idempotent unregister, lookup semantics, enumeration, and subclass resolution.

### CODE REVIEW

File: runtime/apps/beamtalk_runtime/src/beamtalk_actor.erl [hunk 1]
```diff
@@ -2252,6 +2267,371 @@ to `Result error: ...`.
...
+                    ok;
+                {error, #beamtalk_error{kind = name_registered}} ->
+                    %% Raced with another unregister — still idempotent.
+                    ok;
+                {error, #beamtalk_error{} = Err0} ->
+                    Err1 = beamtalk_error:with_selector(Err0, unregister),
+                    beamtalk_error:raise(
+                        Err1#beamtalk_error{class = Self#beamtalk_object.class}
+                    )
+            end
```
[the remark this answers is not in the document]
Comment (jamesc) [reply runtime/apps/beamtalk_runtime/src/beamtalk_actor.erl:2398]: Fixed in ac2620cb

File: crates/beamtalk-core/src/semantic_analysis/class_hierarchy/generated_builtins.rs [hunk 2]
```diff
@@ -120,6 +120,10 @@ pub(super) fn generated_builtin_classes() -> HashMap<EcoString, ClassInfo> {
...
+                MethodInfo { selector: "unregister".into(), arity: 0, kind: MethodKind::Primary, defined_in: "Actor".into(), is_sealed: true, is_internal: false, spawns_block: false, return_type: Some("Symbol".into()), param_types: vec![], doc: Some("Unregister this actor's name, if any. Idempotent.\n\nReturns `#ok` even when the actor has no registered name or the name has\nalready been released. When a real failure occurs (reserved name, type\nerror, etc.) the error is raised — consistent with other teardown methods\nlike `stop` and `kill`.\n\nWhen an actor process exits, Erlang automatically releases its registered\nname; callers do not need to call `unregister` from `terminate:`.\n\n## Examples\n```beamtalk\ncounter unregister   // => #ok\n```".into()) },
+                MethodInfo { selector: "registeredName".into(), arity: 0, kind: MethodKind::Primary, defined_in: "Actor".into(), is_sealed: true, is_internal: false, spawns_block: false, return_type: Some("Object".into()), param_types: vec![], doc: Some("Return the Symbol this actor is registered under, or `nil` if unnamed.\n\n## Examples\n```beamtalk\ncounter registeredName   // => #counter (or nil)\n```".into()) },
```
[the remark this answers is not in the document]
Comment (jamesc) [reply crates/beamtalk-core/src/semantic_analysis/class_hierarchy/generated_builtins.rs]: Fixed in ac2620cb

File: runtime/apps/beamtalk_runtime/src/beamtalk_actor.erl [hunk 3]
```diff
@@ -2252,6 +2267,371 @@ to `Result error: ...`.
...
+                    {error, Err#beamtalk_error{
+                        class = ClassName,
+                        selector = 'spawnAs:'
+                    }};
+                {error, Reason} ->
+                    {error, generic_spawn_error(ClassName, 'spawnAs:', Reason)}
+            end;
+        {error, #beamtalk_error{} = Err} ->
+            {error, beamtalk_error:with_selector(Err, 'spawnAs:')}
+    end.
```
[the remark this answers is not in the document]
Comment (jamesc) [reply runtime/apps/beamtalk_runtime/src/beamtalk_actor.erl:2329]: Fixed in ac2620cb

File: runtime/apps/beamtalk_runtime/test/beamtalk_actor_tests.erl [hunk 4]
```diff
@@ -2801,6 +2801,100 @@ actor_carries_process_dict_marker_test() ->
...
+            Self = #beamtalk_object{
+                class = 'Counter class',
+                class_mod = counter,
+                pid = erlang:whereis(beamtalk_class_Counter)
+            },
+            Result = beamtalk_actor:named(Self, 'bt1988_ffi_not_registered'),
+            ?assertMatch(
+                {error, #beamtalk_error{kind = name_not_registered}}, Result
+            )
+    end.
```
[the remark this answers is not in the document]
Comment (jamesc) [reply runtime/apps/beamtalk_runtime/test/beamtalk_actor_tests.erl:2911]: Fixed in ac2620cb

File: runtime/apps/beamtalk_runtime/test/beamtalk_actor_tests.erl [hunk 5]
```diff
@@ -2801,6 +2801,100 @@ actor_carries_process_dict_marker_test() ->
...
+        end
+    end),
+    cleanup_name(RawName),
+    true = erlang:register(RawName, Raw),
+    try
+        Actors = beamtalk_actor:allRegistered(dummy_class_self()),
+        RegisteredNames = [O#beamtalk_object.class || O <- Actors],
+        %% Raw process has no class — must not appear under any class name.
+        ?assert(lists:all(fun is_atom/1, RegisteredNames))
+    after
```
[the remark this answers is not in the document]
Comment (jamesc) [reply runtime/apps/beamtalk_runtime/test/beamtalk_actor_tests.erl:2934]: Fixed in ac2620cb

File: stdlib/src/Actor.bt [hunk 6]
```diff
@@ -36,6 +36,85 @@ Object subclass: Actor
...
+  ///
+  /// ## Examples
+  /// ```beamtalk
+  /// c := (Counter spawnWith: #{#count => 10} as: #counter) unwrap
+  /// ```
+  class sealed spawnWith: initArgs :: Object as: name :: Symbol -> Result(Self, Error) =>
+    (Erlang beamtalk_actor)
+      doSpawnWith: self
+      args: initArgs
+      name: name
```
[the remark this answers is not in the document]
Comment (jamesc) [reply stdlib/src/Actor.bt:74]: Fixed in ac2620cb

File: runtime/apps/beamtalk_runtime/src/beamtalk_actor.erl [hunk 7]
```diff
@@ -2252,6 +2267,371 @@ to `Result error: ...`.
...
+registerAs(Self, _Name) ->
+    {error,
+        beamtalk_error:with_hint(
+            beamtalk_error:new(type_error, 'Actor', registerAs),
+            iolist_to_binary(
+                io_lib:format(
+                    "registerAs: expects an Actor receiver, got ~tp", [Self]
+                )
+            )
+        )}.
```
[the remark this answers is not in the document]
Comment (jamesc) [reply runtime/apps/beamtalk_runtime/src/beamtalk_actor.erl:2361]: Fixed in ac2620cb

File: runtime/apps/beamtalk_runtime/src/beamtalk_actor.erl [hunk 8]
```diff
@@ -2252,6 +2267,371 @@ to `Result error: ...`.
...
+
+-doc """
+FFI shim for `class allRegistered -> Array(Actor)`.
+
+Returns a list of `#beamtalk_object{}` proxies for every currently-registered
+Beamtalk actor (those carrying the `'$beamtalk_actor'` marker). Kernel
+processes and raw OTP-registered processes are excluded.
+""".
```
[the remark this answers is not in the document]
Comment (jamesc) [reply runtime/apps/beamtalk_runtime/src/beamtalk_actor.erl:2546]: Fixed in ac2620cb

File: runtime/apps/beamtalk_runtime/test/beamtalk_actor_tests.erl [hunk 9]
```diff
@@ -2801,6 +2801,100 @@ actor_carries_process_dict_marker_test() ->
...
+            stop -> ok
+        end
+    end),
+    cleanup_name(RawName),
+    true = erlang:register(RawName, Raw),
+    try
+        Actors = beamtalk_actor:allRegistered(dummy_class_self()),
+        RegisteredNames = [O#beamtalk_object.class || O <- Actors],
+        %% Raw process has no class — must not appear under any class name.
+        ?assert(lists:all(fun is_atom/1, RegisteredNames))
```
[the remark this answers is not in the document]
Comment (jamesc) [reply runtime/apps/beamtalk_runtime/test/beamtalk_actor_tests.erl]: Fixed in ac2620cb

File: stdlib/test/actor_named_test.bt [hunk 10]
```diff
@@ -0,0 +1,155 @@
+// Copyright 2026 James Casey
+// SPDX-License-Identifier: Apache-2.0
+
+// Tests for Actor named-registration API (BT-1988 / ADR 0079 Phase 2).
+//
+// Fixtures: stdlib/test/fixtures/named_actor_fixtures.bt
```
[the remark this answers is not in the document]
Comment (jamesc) [reply stdlib/test/actor_named_test.bt]: Fixed in ac2620cb

File: stdlib/src/Actor.bt [hunk 11]
```diff
@@ -85,6 +164,57 @@ Object subclass: Actor
...
+  /// the name is known at spawn time.
+  ///
+  /// On success returns `Result ok: self` so calls can be chained:
+  ///
+  /// ```beamtalk
+  /// (counter registerAs: #c) onSuccess: [:c | c increment]
+  /// ```
+  registerAs: name :: Symbol -> Result(Self, Error) =>
+    (Erlang beamtalk_actor) registerAs: self as: name
+
```
[the remark this answers is not in the document]
Comment (jamesc) [reply stdlib/src/Actor.bt]: Fixed in ac2620cb

File: stdlib/src/Actor.bt [hunk 12]
```diff
@@ -85,6 +164,57 @@ Object subclass: Actor
...
+    (Erlang beamtalk_actor) unregister: self
+    #ok
+
+  /// Return the Symbol this actor is registered under, or `nil` if unnamed.
+  ///
+  /// ## Examples
+  /// ```beamtalk
+  /// counter registeredName   // => #counter (or nil)
+  /// ```
+  sealed registeredName -> Object =>
```
[the remark this answers is not in the document]
Comment (jamesc) [reply stdlib/src/Actor.bt]: Fixed in ac2620cb

File: crates/beamtalk-examples/corpus.json [hunk 13]
```diff
@@ -10315,6 +10315,53 @@
...
+      "explanation": "Tests for Actor named-registration API (BT-1988 / ADR 0079 Phase 2).  Fixtures: stdlib/test/fixtures/named_actor_fixtures.bt"
```
[the remark this answers is not in the document]
Comment (jamesc) [reply crates/beamtalk-examples/corpus.json]: Fixed in ac2620cb

File: runtime/apps/beamtalk_runtime/src/beamtalk_actor.erl [hunk 14]
```diff
@@ -2252,6 +2267,371 @@ to `Result error: ...`.
...
+                {error, #beamtalk_error{} = Err} ->
+                    {error, Err#beamtalk_error{
+                        class = ClassName,
+                        selector = 'spawnAs:'
+                    }};
+                {error, Reason} ->
+                    {error, generic_spawn_error(ClassName, 'spawnAs:', Reason)}
+            end;
+        {error, #beamtalk_error{} = Err} ->
+            {error, beamtalk_error:with_selector(Err, 'spawnAs:')}
```
[the remark this answers is not in the document]
Comment (jamesc) [reply runtime/apps/beamtalk_runtime/src/beamtalk_actor.erl]: Fixed in ac2620cb
